server/src/main/java/org/opensearch/common/AdjustableSemaphore.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common;
+
+import java.util.concurrent.Semaphore;
+
+/**
+ * AdjustableSemaphore is Extended Semphore where we can change maxPermits.
+ */
+public class AdjustableSemaphore extends Semaphore {
+    private final Object maxPermitsMutex = new Object();
+    private int maxPermits;
+
+    public AdjustableSemaphore(int maxPermits) {
+        super(maxPermits);
+        this.maxPermits = maxPermits;
+    }
+
+    /**
+     * Update the maxPermits in semaphore
+     */
+    public void setMaxPermits(int permits) {
+        synchronized (maxPermitsMutex) {
+            final int diff = Math.subtractExact(permits, maxPermits);
+            if (diff > 0) {
+                // add permits
+                release(diff);
+            } else if (diff < 0) {
+                // remove permits
+                reducePermits(Math.negateExact(diff));
```

Comment:
What is the behavior if permits are already taken ?  Will `releasing` them later result in higher permits ?

Reply:
No for reducePermits case, it shrinks the available permits by given diff. `releasing` already acquired permits wont affect it. 

For more understanding of It, we have added UT of it as well.